CRM/Civisocial/Backend/OAuthProvider/Facebook.php

```diff
@@ -0,0 +1,243 @@
+	 *		to make requests.
+	 */
+	public function __construct($accessToken = NULL) {
+		$this->apiUri = 'https://graph.facebook.com/v2.6';
+		$this->getApiCredentials($this->alias);
+		$this->token = $accessToken;
+	}
+
+	/**
+	 * Authorization URI that user will be redirected to for login
+	 *
+	 * @param array $permissions
+	 *		Permissions to be requested
+	 *
+	 * @return string | bool
+	 * @todo Check if requests have been reviewed by Facebook
+	 */
+	public function getLoginUri($permissions = array(), $reRequest = FALSE) {
+		$uri = 'https://www.facebook.com/dialog/oauth';
+		$params = array(
+			'client_id' => $this->apiKey,
+			'redirect_uri' => $this->getCallbackUri($this->alias),
+		);
+		if (empty($permissions)) {
+			$params['scope'] = implode(',', array_merge($this->getBasicPermissions(), $this->getExtraPermissions()));
+		} else {
+			$params['scope'] = implode(',', $permissions);
+		}
+		if ($reRequest) {
+	  		$params['auth_type'] = 'rerequest';
+		}
+		return $uri."?".http_build_query($params);
+	}
+
+	/**
+	 * Minimum permissions required to use the login
+	 */
+	public function getBasicPermissions() {
+		return array(
+			'public_profile',
+			'email',
+		);
+	}
+
+	/**
+	 * Extra recommended permissions
+	 * 'rsvp_events' and 'publish_actions' require to be reviewed by
+	 * Facebook before the app can request it
+	 */
+	public function getExtraPermissions() {
+		return array(
+			'user_likes',
+			'rsvp_event',
+			'publish_actions',
+		);
+	}
+
+	/**
+	 * Process authentication information returned by OAuth provider after login
+	 */
+	public function handleCallback() {
+		parent::handleCallback();
+		$session = CRM_Core_Session::singleton();
+		$requestOrigin = $session->get("civisocialredirect");
+		if (!$requestOrigin) {
+			$requestOrigin = CRM_Utils_System::url('civicrm', NULL, TRUE);
+		}
+
+		// Facebook sends a code to the callback url, this is further used to acquire
+		// access token from facebook, which is needed to get all the data from facebook
+		if (!isset($_GET['code'])) {
+			exit("Invalid request.");
+		}
+
+		// Make an API request to obtain Access Token
+		// GET params
+		$params = array(
+			'client_id' => $this->apiKey,
+			'client_secret' => $this->apiSecret,
+			'code' => CRM_Utils_Array::value('code', $_GET),
+			'redirect_uri' => $this->getCallbackUri($this->alias),
+		);
+
+		$response = $this->get('oauth/access_token', $params);
+		if (isset($response['error'])) {
+			exit($response['error']);
+		}
+
+		$this->token = CRM_Utils_Array::value('access_token', $response);
+
+		// Check if all basic perimissions have been granted
+		$deniedPermissions = $this->checkPermissions($this->getBasicPermissions());
+
+		if (!empty($deniedPermissions)) {
+			CRM_Utils_System::redirect($this->getLoginUri($deniedPermissions, TRUE));
+			// @todo:	It would be better if we inform first (eg. You need to provide
+			//			email to continue) and then provide a link to re-authorize
+		}
+
```

Comment:
Instead of forcing a user for all the permissions, it might be a nice idea to enable features based on allowed permissions.